pkg/cache/fake.go

```diff
@@ -44,14 +44,37 @@ var _ ServiceAccountCache = &FakeServiceAccountCache{}
 func (f *FakeServiceAccountCache) Start(chan struct{}) {}
 
 // Get gets a service account from the cache
-func (f *FakeServiceAccountCache) Get(name, namespace string) (role, aud string, useRegionalSTS bool, tokenExpiration int64) {
+func (f *FakeServiceAccountCache) Get(name, namespace string) Response {
```

Comment:
nit: It looks like the `Get` and `GetOrNotify` functions are identical here. Would it better to mirror the original `serviceAccountCache` struct, and have `Get` delegate to  `return f.GetOrNotify(name, namespace, nil)`?

Reply:
What exactly do you mean with "mirror the original `serviceAccountCache` struct" here? Just to implement the same behavior? If so, I think this is a good idea, yes. However, I think we missed another aspect while I'm looking over this particular piece of code: I think `GetOrNotify` needs to use a write lock because it modifies the `notificationHandlers` map, while `Get` could work with a read lock. I'm now thinking about using write locks at all places to make it easier and this should not have to much performance impact. And as a consequence of this, we could replace the `sync.RWMutex` with a simpler lock implementation. WDYT?

Reply:
It just occurred to me that we shouldn't need a second mutex, and we can take out a write lock like so:

```
func (c *serviceAccountCache) getSAorNotify(name, namespace string, handler chan struct{}) *Entry {
	c.mu.RLock()
	resp, ok := c.saCache[namespace+"/"+name]
	c.mu.RUnlock()
	if !ok && handler != nil {
		// take out a write lock:
		c.mu.Lock()
		// prevent race condition with adding the handler in "addSA":
		if _, ok := c.notificationHandlers[namespace+"/"+name]; !ok {
			c.notificationHandlers[namespace+"/"+name] = handler
		}
		c.mu.Unlock()
		return nil
	}
	return resp
}
```

Happy to chat if that's easier - you can find me at https://kubernetes.slack.com/

Reply:
Yes, would be nice if it was possible to "elevate" the lock here but it isn't. Nevertheless I now implemented it with a second lock, because:
1. it feels kind of error-prone to not defer the unlocks.
2. the write lock would also block other goroutines which just like to read from other cache regions.